### PR TITLE
 We should only look at Legacy path if DOCKER_CONFIG is not set. 

### PR DESCRIPTION
**Pull request by rhatdan**
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@rhuss PTAL
@vrothberg @mtrmac PTAL

File: pkg/docker/config/config.go [hunk 1]
```diff
@@ -201,13 +201,13 @@ func getAuthFilePaths(sys *types.SystemContext, homeDir string) []authPath {
 			authPath{path: filepath.Join(dockerConfig, "config.json"), legacyFormat: false},
 		)
 	} else {
+		paths = append(paths,
+			authPath{path: filepath.Join(homeDir, dockerLegacyHomePath), legacyFormat: true},
+		)
```
Comment (rhuss) [inline pkg/docker/config/config.go:206]: So is it the last path entry 'wins', right (i.e. the latter overwrites a former entry if both are present in the filesystem). Then the change looks good to me.

Comment (rhatdan) [reply pkg/docker/config/config.go:206]: I actually can not see where last entry wins.  BTW, on a quick look.  It looks to me the first one with a username/password would win.  Although I am not an expert on this code.

**Review by mtrmac — COMMENTED** [review]
> We should only look at Legacy path if DOCKER_HOST is not set.

- That variable is not used
- (That’s **EDIT not** the only thing this PR is changing. Please document your intent.)
- Why? docker/cli AFAICS does not have that condition.

File: pkg/docker/config/config.go (hunk 1 above)
Comment (mtrmac) [reply pkg/docker/config/config.go:206]: Yes, the first entry wins. Which makes this PR hard to reconcile with the original bug report.

**Review by mtrmac — CHANGES_REQUESTED** [review]
[See before]

File: pkg/docker/config/config.go (hunk 1 above)
Comment (mtrmac) [reply pkg/docker/config/config.go:206]:
```go
// getAuthFilePaths returns a slice of authPaths based on the system context
// in the order they should be searched. 
```
and `TestGetAuthPreferNewConfig`.

Comment (rhatdan) [reply pkg/docker/config/config.go:206]: Switched to use legacy last.

**Comment by rhatdan** [comment]
Sorry I meant DOCKER_CONFIG, now fixed commit message.

**Comment by mtrmac** [comment]
The “why” question still remains; this now has no clear relationship to #1290 , and does something else than the original we are trying go be compatible with.

File: pkg/docker/config/config.go (hunk 1 above)
Comment (rhuss) [reply pkg/docker/config/config.go:206]: Sorry that my issue description is so vague.  Unfortunately, I don't have a reproducer anymore, but my problem was solved after I have removed the legacy `~/.dockercfg` that I still had parallel to `~/.docker/config.json`. I will restore it from my backup and will add a more in-depth description of the issue.

**Comment by rhatdan** [comment]
@mtrmac Reworded the PR description as well as the github description.  Even though this does not fix the original problem, it still fixes a potential bug.

**Comment by mtrmac** [comment]
> * Why? docker/cli AFAICS does not have that condition.

> The “why” question still remains; this now has no clear relationship to #1290 , and does something else than the original we are trying go be compatible with.

Why is the current behavior a bug?

**Comment by rhatdan** [comment]
If I set DOCKER_CONFIG and the DOCKER_CONFIG file is not set, but the legacy path is set, then I believe unexpected results happen, if that is ok with you, then the non legacy path should also succeed, and it does not.  We should not treat legacy path special.

**Comment by mtrmac** [comment]
“Unexpected results” by what measure? All of this exists to be compatible to a pre-existing implementation, isn’t that the thing to measure against?

**Comment by rhatdan** [comment]
Unexpected by the user, I would expect to fail if I set DOCKER_CONFIG=/tmp/foobar and it did not exist, but it would succed if the legacy path exists.

But If you don't want this then close it. Most likely would never happen, but I believe this was a mistake when the code was written.

**Comment by mtrmac** [comment]
> Unexpected by the user, I would expect to fail if I set DOCKER_CONFIG=/tmp/foobar and it did not exist, but it would succed if the legacy path exists.

Most importantly, I just don’t think we are in the business of defining of what `DOCKER_CONFIG` should mean. We are reading that to try to do exactly what `/usr/bin/docker` did.

(To be explicit, note that ~/.dockercfg is _not_ in `$DOCKER_CONFIG` which is ~/.docker/`. And taking this thought further, should setting `DOCKER_CONFIG` prevent use of `~/.containers/auth.json`? If not, why is `~/.dockercfg` different?)

> But If you don't want this then close it. Most likely would never happen, but I believe this was a mistake when the code was written.

AFAICS, looking at docker/cli, the current semantics is as expected. If you see some other docker/ code with different behavior, or if I’m reading it incorrectly, please reopen.

**Comment by rhatdan** [comment]
Well then I would say that Docker had broken code...

**Comment by mtrmac** [comment]
That’s not at all obvious to me (`DOCKER_CONFIG` is not _exclusively_ for `config.json`, e.g. the directory is used for TLS secrets of the daemon, `cli-plugins`, Notary, probably more), but even if it were true, I don’t think that’s very relevant: if we aren’t bug-for-bug compatible, we aren’t compatible. Defining a new semantics for `~/.dockercfg` that nothing writes to for the last 6 years (compare https://issues.redhat.com/browse/RUN-1246?focusedCommentId=16344129&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16344129 ) doesn’t help.

**Comment by rhatdan** [comment]
Then we are not compatible since we have fixed other "docker" bugs.

**Comment by mtrmac** [comment]
A good reason to un-fix them, maybe :)

**Comment by rhatdan** [comment]
Nope we will be compatible to a point, but when they are clearly in error, especially ones that modify content of the host outside of the container, then we will not be compatible.

**Pull request closed without merging**